File: SpringTILCode/src/main/java/com/example/springtilcode/domain/user/service/AuthService.java
package com.example.springtilcode.domain.user.service;

import com.example.springtilcode.domain.user.controller.dto.response.TokenResponse;
import com.example.springtilcode.domain.user.entity.User;
import com.example.springtilcode.domain.user.exception.AlreadyExistException;
import com.example.springtilcode.domain.user.exception.PasswordNotMatchException;
import com.example.springtilcode.domain.user.exception.UserNotFoundException;
import com.example.springtilcode.domain.token.RefreshTokenRepository;
import com.example.springtilcode.domain.user.repository.UserRepository;
import com.example.springtilcode.global.error.ErrorCode;
import com.example.springtilcode.global.jwt.TokenProvider;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import javax.transaction.Transactional;

@Service
@RequiredArgsConstructor
public class AuthService {
    private final UserRepository userRepository;
    private final RefreshTokenRepository refreshTokenRepository;
    private final PasswordEncoder passwordEncoder;
    private final TokenProvider tokenProvider;

    @Transactional
    public void signup(String email, String password){
        if(userRepository.findByEmail(email).isPresent())
            throw new AlreadyExistException(ErrorCode.ALREADY_EXIST);
        User user = new User(email, password, passwordEncoder);
        userRepository.save(user);
    }

    @Transactional
    public TokenResponse login(String email, String password){
        User user = userRepository.findByEmail(email).orElseThrow(UserNotFoundException::new);
        if(!passwordEncoder.matches(password, user.getPassword())){
            throw new PasswordNotMatchException();
        }
        String accessToken = tokenProvider.createAccessToken(email);
        String refreshToken = tokenProvider.createRefreshToken(email);

        return TokenResponse.builder()
                .accessToken(accessToken)
                .refreshToken(refreshToken)
                .build();
    }
}

File: kotlin/src/main/kotlin/com/example/demo/domain/user/service/AuthService.kt
package com.example.demo.domain.user.service

import com.example.demo.domain.user.api.dto.response.TokenResponse
import com.example.demo.domain.user.api.dto.response.UserResponse
import com.example.demo.domain.user.domain.Role
import com.example.demo.domain.user.domain.User
import com.example.demo.domain.user.exception.AlreadyExistException
import com.example.demo.domain.user.exception.EmailNotFoundException
import com.example.demo.domain.user.exception.PasswordNotMatchesException
import com.example.demo.domain.user.repository.UserRepository
import com.example.demo.global.error.GlobalErrorCode
import com.example.demo.global.jwt.JwtProvider
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class AuthService (
    private val userRepository: UserRepository,
    private val jwtProvider: JwtProvider,
    private val passwordEncoder: PasswordEncoder
){
    @Transactional
    fun signup(email: String, password: String): UserResponse {
        if(userRepository.findByEmail(email)?.equals(null) == false)
            throw AlreadyExistException(GlobalErrorCode.ALREADY_EXIST)
        val user = User(email, passwordEncoder.encode(password), Role.ROLE_USER)
        userRepository.save(user)
        return UserResponse(user.email, null)
    }

    @Transactional
    fun login(email: String, password: String): TokenResponse {
        val user = userRepository.findByEmail(email) ?: throw EmailNotFoundException(GlobalErrorCode.EMAIL_NOT_FOUND)
        if(!passwordEncoder.matches(password, user.password))
            throw PasswordNotMatchesException(GlobalErrorCode.BAD_REQUEST)

        val access = jwtProvider.createAccessToken(email, user.role)
        val refresh = jwtProvider.createRefreshToken(email)

        return TokenResponse(access, refresh)
    }
}